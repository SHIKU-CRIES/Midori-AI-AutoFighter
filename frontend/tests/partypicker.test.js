import { describe, expect, test } from 'bun:test';
import { readFileSync } from 'fs';
import { join } from 'path';

describe('PartyPicker component', () => {
  test('contains party picker markup', () => {
    const content = readFileSync(join(import.meta.dir, '../src/lib/PartyPicker.svelte'), 'utf8');
    expect(content).toContain('data-testid="party-picker"');
  });

  test('includes add/remove control', () => {
    const content = readFileSync(join(import.meta.dir, '../src/lib/StatTabs.svelte'), 'utf8');
    expect(content).toContain('Add to party');
  });

  test('references updated stat keys', () => {
    const content = readFileSync(join(import.meta.dir, '../src/lib/StatTabs.svelte'), 'utf8');
    expect(content).toContain('crit_damage');
    expect(content).toContain('effect_hit_rate');
    expect(content).toContain('dodge_odds');
    expect(content).toContain('effect_resistance');
  });

  test('filters unowned characters and normalizes element names', () => {
    const content = readFileSync(join(import.meta.dir, '../src/lib/PartyPicker.svelte'), 'utf8');
    expect(content).toContain('filter((p) => p.owned || p.is_player)');
    expect(content).toContain('selected = selected.filter((id) => roster.some((c) => c.id === id))');
    expect(content).toContain('element: resolveElement(p)');
  });

  test('orders stats correctly', () => {
    const content = readFileSync(join(import.meta.dir, '../src/lib/StatTabs.svelte'), 'utf8');
    const coreStart = content.indexOf("{#if activeTab === 'Core'}");
    const coreEnd = content.indexOf("{:else if activeTab === 'Offense'}");
    const coreSection = content.slice(coreStart, coreEnd);
    expect(coreSection).toMatch(/HP[\s\S]*EXP[\s\S]*Vitality[\s\S]*Regain/);
    expect(coreSection).not.toMatch(/DEF/);
    const defStart = content.indexOf("{:else if activeTab === 'Defense'}");
    const defEnd = content.indexOf('{/if}', defStart);
    const defSection = content.slice(defStart, defEnd);
    expect(defSection).toMatch(/<div><span>DEF<\/span><span>{sel.stats.defense/);
  });

  test('uses element colors for icon and outline', () => {
    const rosterContent = readFileSync(join(import.meta.dir, '../src/lib/PartyRoster.svelte'), 'utf8');
    expect(rosterContent).toContain('style={`border-color: ${getElementColor(char.element)}`}');
    expect(rosterContent).toContain('style={`color: ${getElementColor(char.element)}`}');
  });

  test('roster layout snapshot', () => {
    const content = readFileSync(join(import.meta.dir, '../src/lib/PartyRoster.svelte'), 'utf8');
    expect(content).toMatchSnapshot();
  });
});

