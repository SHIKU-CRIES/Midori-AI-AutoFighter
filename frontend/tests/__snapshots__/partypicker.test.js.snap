// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`PartyPicker component roster layout snapshot 1`] = `
"<script>
  import { getElementIcon, getElementColor } from './assetLoader.js';

  /**
   * Displays the available roster and handles character selection.
   *
   * Props:
   * - roster: array of character objects
   * - selected: array of selected character IDs
   * - previewId: ID of the character currently previewed (two-way bound)
   * - compact: render compact icon list when true
   */
  export let roster = [];
  export let selected = [];
  export let previewId;
  export let compact = false;

  function select(id) {
    previewId = id;
  }
</script>

{#if compact}
<div class="roster list compact" data-testid="roster">
  {#each roster.filter(c => selected.includes(c.id)) as char}
    <button
      data-testid={\`choice-\${char.id}\`}
      class="char-btn"
      on:click={() => select(char.id)}>
      <img src={char.img} alt={char.name} class="compact-img" />
    </button>
  {/each}
</div>
{:else}
<div class="roster-list">
  {#each roster as char}
    <button
      type="button"
      data-testid={\`choice-\${char.id}\`}
      class="char-row"
      class:selected={selected.includes(char.id)}
      on:click={() => select(char.id)}
      style={\`border-color: \${getElementColor(char.element)}\`}> 
      <img src={char.img} alt={char.name} class="row-img" />
      <span class="row-name">{char.name}</span>
      <svelte:component
        this={getElementIcon(char.element)}
        class="row-type"
        style={\`color: \${getElementColor(char.element)}\`}
        aria-hidden="true" />
    </button>
  {/each}
</div>
{/if}

<style>
.roster-list {
  display: flex;
  flex-direction: column;
  gap: 0.4rem;
  padding: 0.4rem;
  height: 100%;
  overflow-y: auto;
  border-right: 2px solid #444;
  border-left: 2px solid #444;
  min-width: 0;
}

.char-row {
  display: flex;
  align-items: center;
  gap: 0.5rem;
  padding: 0.25rem 0.4rem;
  background: rgba(0,0,0,0.6);
  border: 2px solid transparent;
  border-radius: 6px;
  cursor: pointer;
  transition: background 0.2s, box-shadow 0.2s;
}
.char-row:hover { background: rgba(20,20,20,0.8); }
.char-row.selected {
  border-color: #ffd700;
  box-shadow: 0 0 8px rgba(255,215,0,0.5);
}
.row-img {
  width: 40px;
  height: 40px;
  object-fit: cover;
  border-radius: 4px;
  border: 1px solid #222;
  flex-shrink: 0;
}
.row-name {
  flex: 1;
  text-align: left;
  color: #fff;
  font-size: 0.9rem;
}
.row-type { width: 20px; height: 20px; flex-shrink: 0; }

/* compact mode */
.roster.list.compact {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 0.25rem;
  padding: 0.25rem;
  background: transparent;
  border: none;
  min-height: 32px;
  height: auto;
}
.roster.list.compact .char-btn {
  border: none;
  background: transparent;
  padding: 0;
  flex-shrink: 0;
}
.roster.list.compact .char-btn img {
  width: 28px;
  height: 28px;
  border-radius: 4px;
  object-fit: cover;
  border: 1px solid #fff;
}
</style>
"
`;
